lib/omnibus/packagers/msi.rb

```diff
@@ -47,24 +47,46 @@ class Packager::MSI < Packager::Base
       FileSyncer.glob("#{resources_path}/assets/*").each do |file|
         copy_file(file, "#{resources_dir}/assets/#{File.basename(file)}")
       end
+
+      # If fastmsi, copy externally built custom action DLL into the staging directory
+      copy_file("#{resources_path}/ext/CustomActionFastMsi.CA.dll", "#{staging_dir}") if fast_msi
     end
 
     build do
+      # If fastmsi, zip up the contents of the install directory
+      zip_command = <<-EOH.split.join(' ').squeeze(' ').strip
+        7z a -r
+        #{windows_safe_path(staging_dir)}\\#{project.name}.zip
+        #{windows_safe_path(project.install_dir)}\\*
+      EOH
+      shellout!(zip_command, returns: [0]) if fast_msi
+
       # Harvest the files with heat.exe, recursively generate fragment for
       # project directory
       Dir.chdir(staging_dir) do
-        shellout! <<-EOH.split.join(' ').squeeze(' ').strip
-          heat.exe dir "#{windows_safe_path(project.install_dir)}"
-            -nologo -srd -sreg -gg -cg ProjectDir
-            -dr PROJECTLOCATION
-            -var "var.ProjectSourceDir"
+        if fast_msi
+          heat_command = <<-EOH.split.join(' ').squeeze(' ').strip
+            heat.exe file "#{project.name}.zip"
+            -cg ProjectDir
+            -dr INSTALLLOCATION
+            -nologo -sfrag -srd -sreg -gg
             -out "project-files.wxs"
-        EOH
+          EOH
+        else
+          heat_command = <<-EOH.split.join(' ').squeeze(' ').strip
+            heat.exe dir "#{windows_safe_path(project.install_dir)}"
+              -nologo -srd -sreg -gg -cg ProjectDir
+              -dr PROJECTLOCATION
+              -var "var.ProjectSourceDir"
+              -out "project-files.wxs"
+          EOH
+        end
+        shellout!(heat_command, returns: [0])
```

Comment:
It might be nice to make `heat_command` a method that returns the relevant command as this build block is getting rather long.